Fix issues inserting Widgets with nested WYSIWYGs

molovo opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR fixes incorrect functionality when inserting widgets which contain a WYSIWYG field.

(Thanks to @EduardTd for pointing me in the right direction).

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#19742: Widgets with a WYSIWYG parameter fail when inserting them into a WYSIWYG in a form. 
2. magento/magento2#13409: custom widget with wysiwyg problem on insert widget via pages or blocks

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
Steps taken are the same used to reproduce #19742 

(Code examples in [this gist](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0))

1. Create a WYSIWYG block as shown in [Wysiwyg.php](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0#file-wysiwyg-php)
2. Create a widget with a parameter which uses the WYSIWYG block as shown in [widget.xml](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0#file-widget-xml)
3. Edit a CMS page, and insert the new widget into the content WYSIWYG.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

molovo committed:
Fix issues inserting Widgets with nested WYSIWYGs

This PR is a potential fix for issues #19742 and #13409 (Thanks to @EduardTd for pointing me in the right direction).

I've tested this briefly in my own install and it appears to resolve the issue. Submitting this PR so that the issue can be tested fully.

magento-engcom-team commented:
Hi @molovo. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=20174) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=20174) before we can accept your contribution.<br/>

VladimirZaets commented on lib/web/mage/adminhtml/wysiwyg/widget.js:
```diff
@@ -513,7 +513,7 @@ define([
          */
         getWysiwyg: function () {
-            return wysiwyg.activeEditor();
+            return tinyMCE.get(this.widgetTargetId);
```
Hi @molovo, thanks for collaboration. 
Where you get `tineMCE` variable? I don't see any declaration of this variable.

molovo replied:
Hi @VladimirZaets - the `tinyMCE` variable is attached to window when TinyMCE is initialised.

Interestingly, it looks like there's actually two instances being loaded in admin, `window.tinyMCE` and `window.tinymce`. Both are the same version.

VladimirZaets replied:
Hi @molovo. 
yes, I know that TinyMCE currently is in global scope ( in window object ). But it's a legacy way, currently, we have some DI that implemented by RequireJS library.  So, we shouldn't rely on the property in the global scope, we should declare dependency directly by RequireJS dependency injection mechanism

molovo replied:
I've fixed this in my latest commit

molovo committed:
Use wysiwygAdapter instead of global tinyMCE variable

I realised that tinyMCE is already being required as the wysiwygAdapter, so I've updated the code to point to that dependency instead. Still works.

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-3953](https://jira.corp.magento.com/browse/ENGCOM-3953) has been created to process this Pull Request

magento-engcom-team commented:
@molovo thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

VasylShvorak commented:
Hi @molovo there are some problems  after testing PR:
**Case 1:**
* Create a WYSIWYG block as shown in [Wysiwyg.php](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0#file-wysiwyg-php)
* Create a widget with a parameter which uses the WYSIWYG block as shown in [widget.xml](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0#file-widget-xml)
* Go to Content->Pages
* Edit Home Page
* Click Content->Insert Widget icon
* Select created widget
* Add some text to wysiwyg field
* Click Insert Widget
* Check Content
:heavy_check_mark: Widget icon is shown
![20174](https://user-images.githubusercontent.com/26908019/53635898-36113800-3c27-11e9-8794-3131f57597c2.png)
* Go to Home Page
:x: Widget content isn't displayed

**Case 2:**
* Create a WYSIWYG block as shown in [Wysiwyg.php](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0#file-wysiwyg-php)
* Create a widget with a parameter which uses the WYSIWYG block as shown in [widget.xml](https://gist.github.com/molovo/051e3882b802fa383a9b1b1b7e4f81c0#file-widget-xml)
* Go to Content->Pages
* Edit Home Page
* Click Content->Insert Widget icon
* Select created widget
* Add image into wysiwyg field
* Click Insert Widget
* Check Content
:x: Broken widget icon is shown instead of inserted image
![20174admin](https://user-images.githubusercontent.com/26908019/53636078-afa92600-3c27-11e9-9b13-76285e8256ee.png)

sidolov commented:
@molovo , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

molovo commented:
@sidolov this PR only addresses the issues with selecting the correct WYSIWYG to insert the widget into.

Regarding the errors described by @VasylShvorak above, I'm unable to reproduce Case 1 - after saving the page and clearing the cache the widget appears correctly on the front end.

Case 2 occurs before applying this patch, and is caused by bugs elsewhere in the code messing up the encoding of the widget snippet which is inserted. IMO this PR should be merged, and that error addressed in a separate PR.

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-4664](https://jira.corp.magento.com/browse/ENGCOM-4664) has been created to process this Pull Request

VasylShvorak commented:
:heavy_check_mark: QA passed
